### PR TITLE
[go_router] Provide return values for replace methods

### PR DESCRIPTION
**Pull request by votruk**
This PR is aimed to add the `Future<T?>` results for the following methods:
* `pushReplacement`
* `pushReplacementNamed`
* `replace`
* `replaceNamed`

This PR addresses the following issue: https://github.com/flutter/flutter/issues/124686

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

PS I'm not sure either the patch or minor version should be changed. Ready to update the PR in case I made something incorrect

### CODE REVIEW

File: packages/go_router/test/delegate_test.dart [hunk 1]
```diff
@@ -179,33 +194,59 @@ void main() {
         );
       },
     );
+
+    testWidgets('gets the value provided from pop',
```
Comment (chunhtai) [inline packages/go_router/test/delegate_test.dart:198]: I think we should move these test to go_router_test and test GoRouter api directly.

**Review by chunhtai — COMMENTED** [review]
change looks good, just one comment on test

File: packages/go_router/test/delegate_test.dart (hunk 1 above)
Comment (votruk) [reply packages/go_router/test/delegate_test.dart:198]: Yeah, true, I was in doubt where to place it. I will make a change 👍

**Comment by chunhtai** [comment]
Hi @votruk are you still planning on coming back to this pr?

**Comment by chunhtai** [comment]
@votruk I will close this pr for now, feel free to reopen if you have time to come back to this pr

**Pull request closed without merging**